handler.go

```diff
@@ -1,13 +1,118 @@
 package metrics
 
 import (
+	"fmt"
 	"net/http"
 
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
 // Handler returns the global http.Handler that provides the prometheus
-// metrics format on GET requests
+// metrics format on GET requests. This handler is no longer instrumented.
 func Handler() http.Handler {
-	return prometheus.Handler()
+	return promhttp.Handler()
+}
+
+// HttpHandlerOpts describes a set of configurable options of http metrics
+type HttpHandlerOpts struct {
+	Namespace          string
+	Subsystem          string
+	HandlerName        string
+	DurationBucket     []float64
+	RequestSizeBucket  []float64
+	ResponseSizeBucket []float64
+}
+
+var (
+	defaultDurationBucket     = []float64{0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1., 2., 5., 10., 20., 30., 40., 50.}
+	defaultRequestSizeBucket  = prometheus.ExponentialBuckets(1024, 2, 22) //1K to 4G
+	defaultResponseSizeBucket = defaultRequestSizeBucket
+)
+
+func InstrumentHandler(handlerName string, handler http.Handler) http.HandlerFunc {
+	return InstrumentHandlerFunc(handlerName, handler.ServeHTTP)
+}
+
+func InstrumentHandlerFunc(handlerName string, handlerFunc func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	return InstrumentHandlerFuncWithOpts(
+		HttpHandlerOpts{
+			Subsystem:   "http",
+			HandlerName: handlerName,
+		},
+		handlerFunc,
+	)
+}
+
+func InstrumentHandlerWithOpts(opts HttpHandlerOpts, handler http.Handler) http.HandlerFunc {
+	return InstrumentHandlerFuncWithOpts(opts, handler.ServeHTTP)
+}
+
+// InstrumentHandlerFuncWithOpts instruments http handler with in-flight gauge,
+// request count, handle duration, request size, and response size
+func InstrumentHandlerFuncWithOpts(opts HttpHandlerOpts, handlerFunc func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	inFlightGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: opts.Namespace,
+		Subsystem: opts.Subsystem,
+		Name:      fmt.Sprintf("%s_in_flight_requests", opts.HandlerName),
```

Comment:
Shouldn't handler names be labels?

Reply:
I agree that using it is better to use labels. However the `promhttp` API only allows "code" and "method" as labels:
https://godoc.org/github.com/prometheus/client_golang/prometheus/promhttp#InstrumentHandlerCounter

Reply:
Label the counter before you pass it to that function. Is that possible?

The issue here is there is nothing to enforce naming standard on the handler name.

Reply:
I can give a try to see if we can label the counter. Well, not only counter, but almost all the APIs only accept "code" and "method" as labels.

If the handler name is not standard, say use "blob-upload" instead of "blob_update" the API will panic at run time.

Reply:
Ok, these are fine then. It would be a lot easier to service-wide rollups if these are labels, but you'd have to rewrite all of these.